### PR TITLE
Optimize projectiles

### PR DESCRIPTION
**Pull request by Garanas**
This pull request will once completed completely optimize all the projectiles in the game. These are the definition of garbage creation as they are allocated just to explode again. Initial tests show that we can create a significant difference in sim time, as an example this [earlier pull request](https://github.com/FAForever/fa/pull/3392).

Work to be done: 
 - [x] /lua/sim/Projectile.lua
 - [x] /lua/sim/DefaultDamage.lua
 - [ ] /lua/sim/DefaultProjectiles.lua
 - [ ] /lua/aeonprojectiles.lua
 - [ ] /lua/cybranprojectiles.lua
 - [ ] /lua/terranprojectiles.lua
 - [ ] /lua/seraphimprojectiles.lua
 - [ ] All individual projectiles of the base game
 - [ ] Tactical missile manager to prevent threads from being made

### CODE REVIEW

File: lua/sim/DefaultDamage.lua [hunk 1]
```diff
@@ -8,21 +8,30 @@
...
+local EntityMethods = _G.moho.entity_methods
+local EntityGetPosition = EntityMethods.GetPosition
+
+-- Applies damage over time to a specific unit.
 UnitDoTThread = function(instigator, unit, pulses, pulseTime, damage, damType, friendly)
     for i = 1, pulses do
-        if unit and not unit:BeenDestroyed() then
-            Damage(instigator, unit:GetPosition(), unit, damage, damType )
-        else
-            break
```
Comment (speed2CZ) [inline lua/sim/DefaultDamage.lua]: If the unit is dead it wont come back to live the next iteration, the break is here for this.

Comment (Garanas) [reply lua/sim/DefaultDamage.lua]: I agree - I'll add it back in

File: lua/sim/Projectile.lua [hunk 2]
```diff
@@ -467,22 +541,59 @@ Projectile = Class(moho.projectile_methods, Entity) {
...
+    -- DEPRECATED --
+
+    -- this should never be called - use the actual function.
+    GetCachePosition = function(self)
+
+        -- deprecation warning in case it ever happens.
+        if not DeprecatedWarnings.CreateCybranBuildBeams then 
+            DeprecatedWarnings.CreateCybranBuildBeams = true 
+            WARN("CreateCybranBuildBeams is deprecated: use projectile:GetPosition() instead.")
+            WARN("Source: " .. repr(debug.getinfo(2)))
```
Comment (Sheikah45) [inline lua/sim/Projectile.lua:567]: Should this be something other than CreateCybranBuildBeams?

File: lua/sim/Projectile.lua [hunk 3]
```diff
@@ -467,22 +541,59 @@ Projectile = Class(moho.projectile_methods, Entity) {
...
+    end,
+
+    -- this should never be called - use the actual value.
+    GetCollideFriendly = function(self)
+
+        -- deprecation warning in case it ever happens.
+        if not DeprecatedWarnings.CreateCybranBuildBeams then 
+            DeprecatedWarnings.CreateCybranBuildBeams = true 
+            WARN("CreateCybranBuildBeams is deprecated: get projectile.CollideFriendly instead.")
+            WARN("Source: " .. repr(debug.getinfo(2)))
```
Comment (Sheikah45) [inline lua/sim/Projectile.lua:580]: Should this be something other than CreateCybranBuildBeams?

File: lua/sim/Projectile.lua [hunk 4]
```diff
@@ -467,22 +541,59 @@ Projectile = Class(moho.projectile_methods, Entity) {
...
+    end,
+
+    -- this should never be called - use the actual value.
+    PassData = function(self, data)
+
+        -- deprecation warning in case it ever happens.
+        if not DeprecatedWarnings.CreateCybranBuildBeams then 
+            DeprecatedWarnings.CreateCybranBuildBeams = true 
+            WARN("CreateCybranBuildBeams is deprecated: set projectile.Data instead.")
+            WARN("Source: " .. repr(debug.getinfo(2)))
```
Comment (Sheikah45) [inline lua/sim/Projectile.lua:593]: Should this be something other than CreateCybranBuildBeams?

File: lua/sim/Projectile.lua (hunk 2 above)
Comment (Garanas) [reply lua/sim/Projectile.lua:567]: Oh - yes!

**Comment by Garanas** [comment]
Moved to milestone 3727 in favor of doing the props first.

**Pull request closed without merging**